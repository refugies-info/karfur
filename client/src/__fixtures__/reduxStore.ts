import { RootState } from "services/rootReducer";
import { SearchResultsState } from "services/SearchResults/searchResults.reducer";
import { UserFavoritesState } from "services/UserFavoritesInLocale/UserFavoritesInLocale.reducer";
import { activeThemesMock } from "./activeThemes";

const initialMockUserState = {
  user: null,
  admin: false,
  traducteur: false,
  expertTrad: false,
  contributeur: false,
  membreStruct: false,
  userId: "",
  userFetched: false,
  rolesInStructure: [],
};
const initialMockLangueState = {
  langues: [],
  languei18nCode: "fr",
  showLanguageModal: false,
  showLangModal: false,
};

const initialMockTranslationState = {
  translation: {
    initialText: {},
    translatedText: {},
  },
  translations: [],
};

const initialUserFavoritesState: UserFavoritesState = {
  favorites: [],
  showFavoriteModal: false
}

const initialMockSearchReults: SearchResultsState = {
  results: {
    dispositifs: [],
    demarches: [],
    dispositifsSecondaryTheme: []
  },
  query: {
    search: "",
    departments: [],
    themes: [],
    needs: [],
    age: [],
    frenchLevel: [],
    language: [],
    sort: "date",
    type: "all",
  }
}

export const initialMockStore = {
  user: initialMockUserState,
  langue: initialMockLangueState,
  activeDispositifs: [],
  tts: { ttsActive: false, showAudioSpinner: false },
  userStructure: null,
  selectedDispositif: null,
  loadingStatus: {},
  translation: initialMockTranslationState,
  activeStructures: [],
  selectedStructure: null,
  allDispositifs: [],
  allStructures: [],
  users: [],
  userFavorites: initialUserFavoritesState,
  userContributions: [],
  needs: [],
  themes: {
    activeThemes: activeThemesMock,
    inactiveThemes: []
  },
  searchResults: initialMockSearchReults
};
