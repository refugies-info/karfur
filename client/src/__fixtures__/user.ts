import { GetUserInfoResponse, UserStatus } from "api-types";
import moment from "moment";

const date = moment("2019-05-01 12:00:00");
export const testUser: GetUserInfoResponse = {
  username: "test",
  _id: "testObjectId",
  structures: [],
  contributions: [],
  phone: "",
  roles: [],
  email: "",
  selectedLanguages: [],
  status: UserStatus.ACTIVE,
  traductionsFaites: []
};

export const testUserWithRoles: GetUserInfoResponse = {
  ...testUser,
  structures: ["testObjectId"],
  roles: [
    { nom: "Admin", _id: "testObjectId", nomPublic: "Admin" },
    { nom: "Trad", _id: "testObjectId", nomPublic: "Trad" },
    {
      nom: "ExpertTrad",
      _id: "testObjectId",
      nomPublic: "ExpertTrad"
    },
    {
      nom: "Contrib",
      _id: "testObjectId",
      nomPublic: "Contrib"
    }
  ]
};
