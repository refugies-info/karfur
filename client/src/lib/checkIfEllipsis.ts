export const checkIfEllipsis = (element: HTMLElement) => element.offsetWidth < element.scrollWidth;
