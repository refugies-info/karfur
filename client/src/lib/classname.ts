export const cls = (...classNames: (string | boolean)[]) => classNames.filter(className => !!className).join(" ");
