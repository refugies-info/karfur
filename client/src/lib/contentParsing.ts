export const getCalloutTranslationKey = (level: "info" | "important") => {
  switch (level) {
    case "info":
      return "callout_info"
    case "important":
      return "callout_important"
  }
}

interface ToParse {
  nodeAttr: string;
  translation: string;
}

export const translationParsing = (originalHTML: string, toParse: ToParse[]) => {
  let parsedHTML = originalHTML;
  for (const parsing of toParse) {
    parsedHTML = parsedHTML.replace(parsing.nodeAttr, `${parsing.nodeAttr} data-title='${parsing.translation}'`)
  }
  return parsedHTML
}
