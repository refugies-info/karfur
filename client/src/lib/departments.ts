import { departments } from "data/departments"

export const getDbDepartment = (depName: string) => {
  const index = departments.map(dep => dep.split(" - ")[1]).indexOf(depName);
  return index > -1 ? departments[index] : depName;
}

/**
 * Get department name or postal code from GMap autocomplete
 * @returns a department number or name formatted
 */
export const formatDepartment = (depName: string) => {
  let depDbFormat = depName;

  // postal code
  if (depName.match(/\d{5}/g)) return depName.slice(0, 2);

  // department
  if (!depDbFormat.includes(" - ")) {
    depDbFormat = getDbDepartment(depName);
  }

  const dep = depDbFormat.split(" - ");
  if (!dep[1]) return dep[0];
  return `${dep[1]} (${dep[0]})`;
}

export const getDepartmentFromNumber = (number: string): string => {
  return departments.find(d => d.startsWith(number)) || "";
}
