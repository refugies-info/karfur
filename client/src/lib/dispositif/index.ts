export * from "./isStatus";
export * from "./getMetadataText";
export * from "./canEdit";
export * from "./canTranslate";
export * from "./getTranslationPageData";
