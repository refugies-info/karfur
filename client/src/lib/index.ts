export const jsUcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toUpperCase() + string.slice(1, string.length);

export const jsLcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toLowerCase() + string.slice(1, string.length);

export const limitNbCaracters = (string: string, nbCaractersMax: number) =>
  string.substring(0, Math.min(string.length, nbCaractersMax)) +
  (string.length > nbCaractersMax ? "..." : "");

export const removeAccents = (str = "") => {
  var accents =
    "ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž";
  var accentsOut =
    "AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz";
  const chars = str.split("");
  let i, x;
  for (i = 0; i < chars.length; i++) {
    if ((x = accents.indexOf(chars[i])) !== -1) {
      chars[i] = accentsOut[x];
    }
  }
  return chars.join("");
};

// rules for password check :
// nbNumUpper = nbNumeric + nbUpper
// score 0 (cannot set password): length < 5
// score 0.5 : 7>length>=5 + nbNumUpper = 1 or 2 or0
// score 1.5 : 7>length >= 5 + nbNumUpper>=3 OR length>=7 + nbNumUpper<3
// score 2.5 : length >= 7 + nbNumUpper>=3
// score 3.5 : length >= 7 + nbNumUpper>=3 + at least 1 special caractere

// make sure to use same function in backend
export const computePasswordStrengthScore = (
  password: string
): { score: number } => {
  if (password.length < 5) {
    return { score: 0 };
  }

  const numerics = password.match(/\d+/g)?.join("");
  const nbNumeric = numerics ? numerics.length : 0;

  const upperCase = password.match(/[A-Z]/g)?.join("");
  const nbUpperCase = upperCase ? upperCase.length : 0;

  if (password.length < 7) {
    if (nbUpperCase + nbNumeric < 3) {
      return { score: 1 };
    }
    return { score: 2 };
  }

  if (nbNumeric + nbUpperCase < 3) {
    return { score: 2 };
  }
  const specialCar = password.match(/[!@#$&*]/g)?.join("");
  const nbSpecialCar = specialCar ? specialCar.length : 0;
  if (nbSpecialCar > 0) {
    return { score: 4 };
  }

  return { score: 3 };
};
