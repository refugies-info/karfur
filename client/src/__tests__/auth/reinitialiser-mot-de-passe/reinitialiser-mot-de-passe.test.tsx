import "jest-styled-components";
import mockRouter from "next-router-mock";
import pageComponent from "pages/auth/reinitialiser-mot-de-passe";
import { initialMockStore } from "__fixtures__/reduxStore";
import { setupGoogleMock } from "__mocks__/react-google-autocomplete";
import { wrapWithProvidersAndRenderForTesting } from "../../../../jest/lib/wrapWithProvidersAndRender";

jest.mock("next/router", () => require("next-router-mock"));

describe("auth/reinitialiser-mot-de-passe", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setupGoogleMock();
  });

  it("renders page if no email", () => {
    const component = wrapWithProvidersAndRenderForTesting({
      Component: pageComponent,
      reduxState: {
        ...initialMockStore,
      },
    });
    expect(component).toMatchSnapshot();
  });
  it("renders page if email", async () => {
    await mockRouter.push("/auth/reinitialiser-mot-de-passe?email=test@example.com");
    const component = wrapWithProvidersAndRenderForTesting({
      Component: pageComponent,
      reduxState: {
        ...initialMockStore,
      },
    });
    expect(component).toMatchSnapshot();
  });
});
