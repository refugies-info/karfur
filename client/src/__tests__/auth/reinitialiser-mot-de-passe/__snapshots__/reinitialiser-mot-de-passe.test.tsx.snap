// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth/reinitialiser-mot-de-passe renders null if no email 1`] = `
<div
  className="container half"
>
  <button
    className="fr-btn fr-btn--tertiary fr-btn--sm fr-icon-arrow-left-line fr-btn--icon-left"
    id="fr-button-:r0:"
    onClick={[Function]}
    style={{}}
  >
    Retour
  </button>
  <div
    className="title"
  >
    <h1>
      Mot de passe oublié
    </h1>
    <p
      className="subtitle"
    >
      Nous allons vous envoyer un mail avec des instructions pour le réinitialiser.
    </p>
  </div>
  <form
    className="container"
    onSubmit={[Function]}
  >
    <div
      className="fr-input-group mb-0"
    >
      <label
        className="fr-label"
        htmlFor="input-:r1:"
      >
        Adresse mail
      </label>
      <input
        aria-describedby="input-:r1:-desc-error"
        autoFocus={true}
        className="fr-input"
        id="input-:r1:"
        name="email"
        type="email"
      />
    </div>
    <div
      className="text-end"
    >
      <button
        className="fr-btn fr-icon-mail-line fr-btn--icon-right button full mt"
        id="fr-button-:r2:"
        style={{}}
        type="submit"
      >
        Envoyer le lien de réinitialisation
      </button>
    </div>
  </form>
</div>
`;

exports[`auth/reinitialiser-mot-de-passe renders page if email 1`] = `
<div
  className="container half"
>
  <button
    className="fr-btn fr-btn--tertiary fr-btn--sm fr-icon-arrow-left-line fr-btn--icon-left"
    id="fr-button-:r3:"
    onClick={[Function]}
    style={{}}
  >
    Retour
  </button>
  <div
    className="title"
  >
    <h1>
      Mot de passe oublié
    </h1>
    <p
      className="subtitle"
    >
      Nous allons vous envoyer un mail avec des instructions pour le réinitialiser.
    </p>
  </div>
  <form
    className="container"
    onSubmit={[Function]}
  >
    <div
      className="fr-input-group mb-0"
    >
      <label
        className="fr-label"
        htmlFor="input-:r4:"
      >
        Adresse mail
      </label>
      <input
        aria-describedby="input-:r4:-desc-error"
        autoFocus={true}
        className="fr-input"
        defaultValue="test@example.com"
        id="input-:r4:"
        name="email"
        type="email"
      />
    </div>
    <div
      className="text-end"
    >
      <button
        className="fr-btn fr-icon-mail-line fr-btn--icon-right button full mt"
        id="fr-button-:r5:"
        style={{}}
        type="submit"
      >
        Envoyer le lien de réinitialisation
      </button>
    </div>
  </form>
</div>
`;
